src/registrar_agent.rs

```diff
@@ -0,0 +1,144 @@
+use anyhow::{anyhow, Result};
+use log::*;
+use reqwest::header::*;
+use serde::{Deserialize, Serialize};
+use serde_json::Number;
+
+fn serialize_as_base64<S>(
+    bytes: &[u8],
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: serde::Serializer,
+{
+    serializer.serialize_str(&base64::encode(bytes))
+}
+
+fn deserialize_as_base64<'de, D>(deserializer: D) -> Result<Vec<u8>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    String::deserialize(deserializer).and_then(|string| {
+        base64::decode(&string).map_err(serde::de::Error::custom)
+    })
+}
+
+fn deserialize_maybe_base64<'de, D>(
+    deserializer: D,
+) -> Result<Option<Vec<u8>>, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    Option::<WrappedBase64Encoded>::deserialize(deserializer)
+        .map(|wrapped| wrapped.map(|wrapped| wrapped.0))
+}
+
+fn is_empty(buf: &[u8]) -> bool {
+    buf.is_empty()
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Register<'a> {
+    #[serde(
+        serialize_with = "serialize_as_base64",
+        skip_serializing_if = "is_empty"
+    )]
+    ekcert: &'a [u8],
+    #[serde(
+        serialize_with = "serialize_as_base64",
+        skip_serializing_if = "is_empty"
+    )]
+    ek_tpm: &'a [u8],
+    #[serde(serialize_with = "serialize_as_base64")]
+    aik_tpm: &'a [u8],
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct RegisterResponseResults {
+    #[serde(deserialize_with = "deserialize_maybe_base64")]
+    blob: Option<Vec<u8>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Activate<'a> {
+    auth_tag: &'a str,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ActivateResponseResults {}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Response<T> {
+    code: Number,
+    status: String,
+    results: T,
+}
+
+#[derive(Debug, Deserialize)]
+struct WrappedBase64Encoded(
+    #[serde(deserialize_with = "deserialize_as_base64")] Vec<u8>,
+);
+
+pub(crate) async fn do_activate_agent(
+    registrar_ip: &str,
+    registrar_port: &str,
+    agent_uuid: &str,
+    auth_tag: &str,
+) -> Result<()> {
+    let data = Activate { auth_tag };
+
+    let addr = format!(
+        "http://{}:{}/agents/{}/activate",
+        registrar_ip, registrar_port, agent_uuid
+    );
+    let resp = reqwest::Client::new().put(&addr).json(&data).send().await?;
+
+    if !resp.status().is_success() {
+        return Err(anyhow!(
```

Comment:
Maybe it makes sense to use one of our custom error types here instead of `anyhow`? I could see a case for either.

Reply:
I ended up using a custom type so that we can check the status codes.